File: test_vulnerable_contract.go
package main

import (
	"fmt"
)

type VulnerableContract struct {
	balances map[string]uint64
	owner    string
}

// Missing error handling
func (vc *VulnerableContract) GetBalance(addr string) uint64 {
	balance, _ := vc.balances[addr] // Ignored error
	return balance
}

// Panic instead of error return
func (vc *VulnerableContract) PanicFunction() {
	panic("Something went wrong") // VULNERABLE: using panic
}

// Missing message validation
func (vc *VulnerableContract) HandleTransferMsg(msg TransferMsg) {
	// No ValidateBasic() call
	// No input validation
	vc.balances[msg.From] -= msg.Amount
	vc.balances[msg.To] += msg.Amount
}

// Missing authorization check
func (vc *VulnerableContract) AdminFunction() {
	// No authorization check - anyone can call this!
	vc.owner = "new_owner"
}

// Unvalidated state transition
func (vc *VulnerableContract) SetState(newState string) {
	// No state validation
	vc.owner = newState
}

// Gas limit issue - unbounded loop
func (vc *VulnerableContract) ProcessLargeArray(arr []uint64) {
	for i := 0; i < len(arr); i++ {
		// Unbounded loop - can run out of gas
		vc.balances["user"] += arr[i]
	}
}

// Array access without bounds checking
func (vc *VulnerableContract) UnsafeArrayAccess(arr []string, index int) string {
	return arr[index] // VULNERABLE: can panic on out-of-bounds
}

// Missing permission check
func (vc *VulnerableContract) GovernanceFunction() {
	// Should check for governance authority
	// But doesn't - VULNERABLE!
	vc.balances = make(map[string]uint64)
}

type TransferMsg struct {
	From   string
	To     string
	Amount uint64
}

// Missing ValidateBasic implementation
func (msg TransferMsg) ValidateBasic() error {
	// Should validate addresses and amounts
	// But implementation is missing
	return nil
}

// Keeper without proper authorization
type VulnerableKeeper struct {
	storeKey string
}

func (k VulnerableKeeper) SetBalance(addr string, amount uint64) {
	// No authorization check - VULNERABLE!
	fmt.Printf("Setting balance for %s to %d without authorization\n", addr, amount)
}

func (k VulnerableKeeper) DeleteAccount(addr string) {
	// No authorization check - VULNERABLE!
	fmt.Printf("Deleting account %s without authorization\n", addr)
}

// Message handler without validation
func HandleVulnerableMsg(msg VulnerableMsg) error {
	// No input validation
	// No authorization check
	// Direct state modification
	fmt.Printf("Handling message from %s without validation\n", msg.Sender)
	return nil
}

type VulnerableMsg struct {
	Sender string
	Data   []byte
}

func (msg VulnerableMsg) GetSigners() []string {
	// Should return proper signers but doesn't validate
	return []string{}
}

func main() {
	contract := &VulnerableContract{
		balances: make(map[string]uint64),
		owner:    "initial_owner",
	}

	// Demonstrate vulnerabilities
	contract.PanicFunction() // Will panic
}

File: test_vulnerable_contract.rs
#![cfg_attr(not(feature = "std"), no_std)]

use ink_lang as ink;

#[ink::contract]
mod vulnerable_contract {
    use ink_storage::traits::SpreadAllocate;
    use ink_storage::Mapping;

    #[ink(storage)]
    #[derive(SpreadAllocate)]
    pub struct VulnerableContract {
        balances: Mapping<AccountId, Balance>,
        owner: AccountId,
        total_supply: Balance,
    }

    impl VulnerableContract {
        #[ink(constructor)]
        pub fn new() -> Self {
            ink_lang::utils::initialize_contract(|contract: &mut Self| {
                let caller = Self::env().caller();
                contract.owner = caller;
                contract.total_supply = 1000000;
                // Missing proper initialization of balances
            })
        }

        // Panic condition - unwrap without checking
        #[ink(message)]
        pub fn get_balance(&self, account: AccountId) -> Balance {
            self.balances.get(&account).unwrap() // VULNERABLE: can panic!
        }

        // Integer arithmetic without checks
        #[ink(message)]
        pub fn unsafe_add(&self, a: u128, b: u128) -> u128 {
            a + b // VULNERABLE: can overflow
        }

        // Unsafe code block
        #[ink(message)]
        pub fn unsafe_operation(&mut self) {
            unsafe {
                // VULNERABLE: unsafe code without proper validation
                let ptr = &mut self.total_supply as *mut Balance;
                *ptr = 0;
            }
        }

        // Direct storage access without validation
        #[ink(message)]
        pub fn direct_balance_set(&mut self, account: AccountId, amount: Balance) {
            self.balances.insert(&account, &amount); // VULNERABLE: no validation
        }

        // Cross-contract call without error handling
        #[ink(message)]
        pub fn call_external(&self, contract_addr: AccountId) {
            let result = ink_env::call::build_call::<ink_env::DefaultEnvironment>()
                .call_type(ink_env::call::Call::new().callee(contract_addr))
                .exec(); // VULNERABLE: result not checked
        }

        // Array indexing without bounds check
        #[ink(message)]
        pub fn unsafe_array_access(&self, arr: Vec<u32>, index: usize) -> u32 {
            arr[index] // VULNERABLE: can panic on out-of-bounds
        }

        // Assert without proper error handling
        #[ink(message)]
        pub fn unsafe_assert(&self, condition: bool) {
            assert!(condition); // VULNERABLE: can panic
        }

        // Expect with potential panic
        #[ink(message)]
        pub fn unsafe_expect(&self, option: Option<u32>) -> u32 {
            option.expect("Value should exist") // VULNERABLE: can panic
        }
    }
}
